File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        for (int i = 0; i < obstacleGrid.length; i++) {
            if (obstacleGrid[i][0] == 1) {
                break;
            }
            
            obstacleGrid[i][0] = -1;
        }
        
        for (int j = 0; j < obstacleGrid[0].length; j++) {
            if (obstacleGrid[0][j] == 1) {
                break;
            }
            
            obstacleGrid[0][j] = -1;
        }
        
        for (int i = 1; i < obstacleGrid.length; i++) {
            for (int j = 1; j < obstacleGrid[0].length; j++) {
                if (obstacleGrid[i][j] == 1) {
                    continue;
                }
                
                int wayY = obstacleGrid[i - 1][j] == 1 ? 0 : obstacleGrid[i - 1][j];
                int wayX = obstacleGrid[i][j - 1] == 1 ? 0 : obstacleGrid[i][j - 1];
                
                obstacleGrid[i][j] = wayY + wayX;
            }
        }
        
        int sizeX = obstacleGrid.length;
        int sizeY = obstacleGrid[0].length;
        int result = obstacleGrid[sizeX - 1][sizeY - 1] * -1;
        if (result == -1) {
            return 0;
        }
        
        return result;
    }
}

File: 0063-unique-paths-ii/0063-unique-paths-ii.go
func uniquePathsWithObstacles(obstacleGrid [][]int) int {
    sizeX := len(obstacleGrid)
    sizeY := len(obstacleGrid[0])
    
    for i := 0; i < sizeX; i++ {
        if obstacleGrid[i][0] == 1 {
            break
        }
        
        obstacleGrid[i][0] = -1
    }
    
    for j := 0; j < sizeY; j++ {
        if obstacleGrid[0][j] == 1 {
            break
        }
        
        obstacleGrid[0][j] = -1
    }
    
    for i := 1; i < sizeX; i++ {
        for j := 1; j < sizeY; j++ {
            if obstacleGrid[i][j] == 1 {
                continue
            }
            
            wayX := 0
            wayY := 0
            if obstacleGrid[i][j - 1] != 1 {
                wayX = obstacleGrid[i][j - 1]
            }
            
            if obstacleGrid[i - 1][j] != 1 {
                wayY = obstacleGrid[i - 1][j]
            }
            
            obstacleGrid[i][j] = wayX + wayY
        }
    }
    
    result := 0
    if obstacleGrid[sizeX - 1][sizeY - 1] != 1 {
        result = obstacleGrid[sizeX - 1][sizeY - 1] * -1
    }
    
    return result
}